### ISSUE TITLE
Expose all storage operations from unit of work

### ISSUE THREAD

**Issue by barclayadam** [JasperFx/marten#1209]
I have been looking at using listeners as a way of implementing more automatic handling of tenant identifiers.

To do that I wanted to check out all the operations before save to potentially set the `TenantOverride` property of the operations (i.e. `DocumentStorageOperation`).

The current API for `IUnitOfWork` does not expose what I need so I ended up with:

````c#
var _operations = (Ref<ImHashMap<Type, IList<IStorageOperation>>>) typeof(UnitOfWork)
    .GetField("_operations", BindingFlags.Instance | BindingFlags.NonPublic)
    .GetValue(session.PendingChanges);

foreach (var op in _operations.Value.Value)
{
    if (op is DocumentStorageOperation storageOperation)
    {
        if (storageOperation.Document is ITenantEntity tenant)
        {
            storageOperation.TenantOverride = tenant.CompanyId.ToString();
        }
    }
}
````

Is it sensible to expose the operations from unit of work to avoid that first bit of reflection?

**Comment by oskardudycz** [comment]
@barclayadam just to confirm the use case: you'd like to be able to save documents having multiple tenants in the single document session?

**Comment by barclayadam** [comment]
TLDR We want automatic setting of the tenantid to avoid potentially saving a document without one and causing querying issues that easily go undetected.

Not quite, no. We are converting a large app away from table store to Postgres + Marten. We want to take advantage of tenancy of the documents.

We have a few apps, one where we can easily use a session with a default tenant, and two others that cannot. For example a background task processor that could potentially modify multiple tenants in one "request".

So that means we would have to, for every entity, change from `session.Store(entity)` to `session.Store(entity.CompanyId, entity)`. Because it's not a problem to use the first method it is _very_ easy to miss and at that point, no tenant id is set and therefore `*DEFAULT*` is stored and means this entity is now not loadable using a tenant-scoped session.

A (potentially) better way of handling this is for Marten to be able to be configured with the property / expression that marks a tenant and it uses that:

````c#
For<ContactEntity>()
     // Company always becomes the tenant id for ContactEntity when saved
    .MultiTenanted(x => x.Company)
````

**Comment by oskardudycz** [comment]
This is something that is on my TODO list. I fully understand that it's really needed, although it won't be so easy to implement to make it right (especially keeping in mind that it has to be also supported for Event Store). 
I don't think that exposing all storage operations would be good design. Personally I'd prefer option that you suggested in the second post (so the making it on the document/event type level).

@jeremydmiller do you have some suggestions or some potential solution?

**Comment by barclayadam** [comment]
The unit of work (from `IDocumentSession.PendingChanges`) does currently expose Inserts/Updates/Patches/Deletes but the design there feels inconsistent. For example, `Inserts` and `Updates` return the actual object that is being stored, `Patches` returns an operation and Deletes returns an `IDeletion`:

![image](https://user-images.githubusercontent.com/973909/53755907-a9c57600-3eaf-11e9-8a32-e9e71cbf8a03.png)

I would really like to see this done properly in the long term, but in the short term consolidation of the methods on `IUnitOfWork` to expose the operations for inserts and updates like is currently done for deletes and patches would provide me what I need without reflection hacks.

**Comment by jeremydmiller** [comment]
The deletions aren't always 1 to 1 with a specific document and patches are also not 1 to 1 with a particular document.  I don't really think there's anything to change. I suppose if you really wanted to, we could implement a "preview what documents would be patched or deleted" by querying the store to see what documents would be impacted. 

I wouldn't make this any kind of high priority

**Comment by barclayadam** [comment]
Thanks @jeremydmiller. I'm not after any real new functionality here, just exposing the list of _all_ storage operations that currently exist.

I can get deletions and patches but want access to upserts, inserts and updates to manipulate.

My suggestion is a single IEnumerable StorageOperations method, or perhaps 3 for each type as above, added to IUnitOfWork to expose the information already available.

Would you accept a PR?

**Comment by jeremydmiller** [comment]
"Would you accept a PR?" -- yep. Always happy to take usability improvements;)

**Comment by barclayadam** [comment]
Ok great. I'll try and get something together next day or so. Cheers!

**Comment by mdissel** [comment]
Maybe row level security could also help to make sure the save actions are for the correct tenant (for an example see https://www.citusdata.com/blog/2018/04/04/raw-sql-access-with-row-level-security/), (currently this is not possible with Marten)

**Comment by oskardudycz** [comment]
@jeremydmiller I think that what @barclayadam wants is not the preview, but possibility to make operations on multiple documents on multiple tenants. Currently you have only possibility to set tenant per document session. Which is imho big limitation.

I agree with @barclayadam that it would be good to have it. Of course, it's not the highest priority right now, but imho we should at least discuss that and potential design.

@mdissel yeah, when I read your link that you send to gitter, same thing came to my mind. We should at least consider and poc that.

Then only issue with all of that is as always - capacity and priorities... If you could work on some of those issues, we'll gladly take help and PRs ;)

**Comment by oskardudycz** [comment]
Fix will be available in Marten 3.5.0
